odl/tomo/data/uncompr_bin.py

```diff
@@ -0,0 +1,716 @@
+
+        Parameters
+        ----------
+        dstart : int, optional
+            Offset in bytes of the data field. By default, it is taken to
+            be the header size as determined from reading the header.
+            Backwards indexing with negative values is also supported.
+            Use a value larger than the header size to extract a data subset.
+        dend : int, optional
+            End position in bytes until which data is read (exclusive).
+            Backwards indexing with negative values  is also supported.
+            Use a value different from the file size to extract a data subset.
+        reshape_order : {'C', 'F', 'A'}, optional
+            Value passed as ``order`` parameter to `numpy.reshape`.
+            Reshaping is only done in case the whole data block is read.
+
+        Returns
+        -------
+        data : `numpy.ndarray`
+            The data read from `file`.
+
+        See Also
+        --------
+        read_header
+        """
+        self.file.seek(0, 2)  # 2 means "from the end"
+        filesize_bytes = self.file.tell()
+        if dstart is None:
+            dstart_abs = int(self.header_size)
+        elif dstart < 0:
+            dstart_abs = filesize_bytes + int(dstart)
+        else:
+            dstart_abs = int(dstart)
+
+        if dend is None:
+            dend_abs = int(filesize_bytes)
+        elif dend < 0:
+            dend_abs = int(dend) + filesize_bytes
+        else:
+            dend_abs = int(dend)
+
+        if dstart_abs >= dend_abs:
+            raise ValueError('invalid `dstart` and `dend`, resulting in '
+                             'absolute `dstart` >= `dend` ({} >= {})'
+                             ''.format(dstart_abs, dend_abs))
+        if dstart_abs < self.header_size:
+            raise ValueError('invalid `dstart`, resulting in absolute '
+                             '`dstart` < `header_size` ({} < {})'
+                             ''.format(dstart_abs, self.header_size))
+        if dend_abs > filesize_bytes:
+            raise ValueError('invalid `dend`, resulting in absolute '
+                             '`dend` > `filesize_bytes` ({} < {})'
+                             ''.format(dend_abs, filesize_bytes))
+
+        num_elems = (dend_abs - dstart_abs) / self.data_dtype.itemsize
+        if num_elems != int(num_elems):
+            raise ValueError('trying to read {} bytes, corresponding to '
+                             '{} elements of type {}'
+                             ''.format(dend_abs - dstart_abs, num_elems,
+                                       self.data_dtype))
+        self.file.seek(dstart_abs)
+        # Numpy determines byte order by itself
+        data = np.fromfile(self.file, dtype=self.data_dtype,
+                           count=int(num_elems))
+
+        if dstart_abs == self.header_size and dend_abs == filesize_bytes:
+            # Full dataset read, reshape to stored shape.
```

Comment:
Well, another option is to not reshape at all. Obviously reshaping is not possible if you read an arbitrary chunk of data. On the other hand, if the header specifies a shape of the data, you probably want your data to be in that shape after reading it from the file.
An alternative would be a boolean parameter reshape yes/no, but that feels a bit redundant and won't play well with the chunk reading situation (ignore? raise exception?).
Yet another alternative would be to never reshape and let the user do it. But then, as usual, they can get the order wrong, and when axis ordering enters the game, things get really awful.

So I would very much want to offer the possibility to reliably get the data reshaped and transposed correctly from the information in the header. It's not so relevant here in the base class, but in MRC it definitely is.